docs/web-apps/serverless/architectures/web-app-content.md

```diff
@@ -265,6 +265,10 @@ If you make a breaking change in an API, publish a new version in API Management
 
 For updates that aren't breaking API changes, deploy the new version to a staging slot in the same Function App. Verify the deployment succeeded and then swap the staged version with the production version. Publish a revision in API Management.
 
+#### Testing
+
+In serverless applications, the line between infrastructure and business logic is blurred. And due to their distributed nature, typically pieces of code are spread across various services. To get the most value-to-effort ratio from testing your serverless applications (especially functions), we recommend you to emphasize more on integration tests and do unit tests only if you have genuinely complex business logic.
+
```

Comment:
Thanks for the frank comment. Let me clarify this further:

Regarding the "_In serverless applications, the line between infrastructure and business logic is blurred._" This sentence was my attempt to prepare the readers for my following sentence (so we can remove it if you think it doesn't serve the purpose). And this sentence, in my opinion, was the gist of conversation between two thought leaders in serverless world. Please look at this paragraph:

> “I think infrastructure-from-code is wrong on all counts. I don’t think we should call it infrastructure… (skipped)… When we talk about business logic, we missed that business logic exists in things that are not general-purpose programming languages. The choice to have an SQS queue in your architecture is a business logic decision. It’s saying that I need persistence for this message that is passing through… (skipped)… Similarly, I’d like to say that least privileged policies is business logic because the choice of what matters and doesn’t matter in an authorization decision, is specific to the business case you’re doing… which is the definition of business logic — the irreducible complexities that are specific to the problems that you’re solving.”

I'm pretty sure the author has explained that sentence much better that me in his [blog post](https://medium.com/theburningmonk-com/are-we-getting-infrastructure-all-wrong-in-the-serverless-era-67e2db07a086). Feel free to check it if it interests you.

Regarding "_nor how their distributed nature means code is spread across "services" (what service are we talking about here?)_". I specifically had Azure Functions in mind. Put it simply: because you can have tens of functions and each function can / should typically have small code and remain single-responsible, considering the effort put on testing and expected outcome, it's probably not wise to do many unit tests because unit testing the business logic of a single function in such a distributed system takes time and effort but doesn't add much value; instead those functions should be mainly tested to check whether they are working well together or with their integrated services. If their integration tests succeed it means that their business logic also works well and vice versa. (so no need to do unit tests if code is small and simple). Even though I had Azure Functions in mind but this can apply to other serverless services too wherein code is scattered and spread across different units (unit can refer to containers, another api, etc) and testing integration is more important than individually unit testing the logic of any of those pieces of code. This [blog post](https://medium.com/theburningmonk-com/my-testing-strategy-for-serverless-applications-636de3d46b6a) has explained it better than me so I recommend you check it

Regarding "_not sure how serverless has such a important role to play in your choice of unit tests or integration tests"_

Well, serverless is a new paradigm and as I explained, they require a different testing approach than traditional testing approach, wherein it's recommended that most tests should be unit tests [Google official blog post](https://testing.googleblog.com/2015/04/just-say-no-to-more-end-to-end-tests.html) is one of the sources that suggests this

_"Can you try to rewrite this?"_ Do you agree with any the aforementioned points, to begin with? If so, I can try to rewrite it. Else, I will close this PR.

Reply:
Thank you. I really like how this conversation is going. It's insightful. I'll also share link to those blog post authors so hopefully they can comment.

Seems we both agree that there is need for more integration tests. I've reduced and modified my proposed text accordingly Also, DevOps is applicable to both frontend and backend but somehow it was mentioned only in frontend. I tried to re-arrange the paragraphs in order to have minimum amount of text duplication.